plumbing/format/pktline/pktline.go

```diff
@@ -0,0 +1,229 @@
+package pktline
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/go-git/go-git/v5/utils/ioutil"
+	"github.com/go-git/go-git/v5/utils/trace"
+)
+
+// WritePacket writes a pktline packet.
+func WritePacket(w io.Writer, p []byte) (n int, err error) {
+	if w == nil {
+		return 0, ErrNilWriter
+	}
+
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: > %04x %s", n, p)
+		}
+	}()
+
+	if len(p) == 0 {
+		return w.Write(emptyPkt)
+	}
+
+	if len(p) > MaxPayloadSize {
+		return 0, ErrPayloadTooLong
+	}
+
+	pktlen := len(p) + lenSize
+	n, err = w.Write(asciiHex16(pktlen))
+	if err != nil {
+		return
+	}
+
+	n2, err := w.Write(p)
+	n += n2
+	return
+}
+
+// WritePacketf writes a pktline packet from a format string.
+func WritePacketf(w io.Writer, format string, a ...interface{}) (n int, err error) {
+	if len(a) == 0 {
+		return WritePacket(w, []byte(format))
+	}
+	return WritePacket(w, []byte(fmt.Sprintf(format, a...)))
+}
+
+// WritePacketln writes a pktline packet from a string and appends a newline.
+func WritePacketln(w io.Writer, s string) (n int, err error) {
+	return WritePacket(w, []byte(s+"\n"))
+}
+
+// WritePacketString writes a pktline packet from a string.
+func WritePacketString(w io.Writer, s string) (n int, err error) {
+	return WritePacket(w, []byte(s))
+}
+
+// WriteErrorPacket writes an error packet.
+func WriteErrorPacket(w io.Writer, e error) (n int, err error) {
+	return WritePacketf(w, "%s%s\n", errPrefix, e.Error())
+}
+
+// WriteFlush writes a flush packet.
+// This always writes 4 bytes.
+func WriteFlush(w io.Writer) (err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: > 0000")
+		}
+	}()
+
+	_, err = w.Write(flushPkt)
+	return err
+}
+
+// WriteDelim writes a delimiter packet.
+// This always writes 4 bytes.
+func WriteDelim(w io.Writer) (err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: > 0001")
+		}
+	}()
+
+	_, err = w.Write(delimPkt)
+	return err
+}
+
+// WriteResponseEnd writes a response-end packet.
+// This always writes 4 bytes.
+func WriteResponseEnd(w io.Writer) (err error) {
+	defer func() {
+		if err == nil {
+			trace.Packet.Printf("packet: > 0002")
+		}
+	}()
+
+	_, err = w.Write(responseEndPkt)
+	return err
+}
+
+// ReadPacket reads a pktline packet.
+// This returns the length of the packet, the packet payload, and an error.
+// The error can be of type *ErrorLine if the packet is an error packet.
+// Use packet length to determine the type of packet i.e. 0 is a flush packet,
+// 1 is a delim packet, 2 is a response-end packet, and a length greater or
+// equal to 4 is a data packet.
+func ReadPacket(r io.Reader) (l int, p []byte, err error) {
```

Comment:
As `ReadPacket` is static we have "limited" options on avoiding unecessary allocations. Currently it has two sources for allocation: `pktlen` and `data`. 

An option here would be able to provide a buffer and rely on a `io.LimitedReader` instead of `io.Reader`. Also adding the trace "fix":
```go
func ReadPacketBuf(lr *io.LimitedReader, buf *bytes.Buffer) (l int, p []byte, err error) {
	defer func() {
		if err == nil {
			trace.Packet.Printfunc(func() string { return fmt.Sprintf("packet: < %04x %s", l, p) })
		}
	}()

	lr.N = 4
	buf.Reset()
	buf.Grow(4)

	n, err := io.Copy(buf, lr)
	if err != nil {
		if errors.Is(err, io.ErrUnexpectedEOF) {
			return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
		}

		return Err, nil, err
	}

	if n != lenSize {
		return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
	}

	length, err := ParseLength(buf.Bytes())
	if err != nil {
		return Err, nil, err
	}

	switch length {
	case Flush, Delim, ResponseEnd:
		return length, nil, nil
	case lenSize: // empty line
		return length, []byte{}, nil
	}

	buf.Reset()
	dataLen := length - lenSize
	buf.Grow(dataLen)

	lr.N = int64(dataLen)
	dn, err := io.Copy(buf, lr)
	if err != nil {
		return Err, nil, err
	}

	if dn != int64(dataLen) {
		return Err, nil, fmt.Errorf("%w: %d", ErrInvalidPktLen, dn)
	}

	data := buf.Bytes()
	if bytes.HasPrefix(data, errPrefix) {
		err = &ErrorLine{
			Text: string(bytes.TrimSpace(data[4:])),
		}
	}

	return length, data, err
}
```

That results in no allocations:
```
BenchmarkReadPacketBuf/empty-16         	1000000000	         0.0000042 ns/op	       0 B/op	       0 allocs/op
BenchmarkReadPacketBuf/one_message-16   	16635014	        74.15 ns/op	       0 B/op	       0 allocs/op
BenchmarkReadPacketBuf/two_messages-16  	16032168	        72.44 ns/op	       0 B/op	       0 allocs/op
BenchmarkReadPacketBuf/sections-16      	16782122	        75.37 ns/op	       0 B/op	       0 allocs/op
```

Reply:
Thanks for the review @pjbgf. I think you're onto something here, we could improve this and make it more generic by using `ReadPacketBuf(io.Reader, []byte)` instead which makes it closer to the upstream `int packet_read` method signature.

Take a look at this implementation:
```go

func ReadPacketBuf(r io.Reader, p []byte) (l int, err error) {
	defer func() {
		if err == nil {
			trace.Packet.Printfunc(func() string { return fmt.Sprintf("packet: < %04x %s", l, p) })
		}
	}()

	if len(p) < lenSize {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, len(p))
	}

	n, err := r.Read(p[:lenSize])
	if err != nil {
		return Err, err
	}

	if n != lenSize {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, n)
	}

	length, err := ParseLength(p)
	if err != nil {
		return Err, err
	}

	switch length {
	case Flush, Delim, ResponseEnd:
		return length, nil
	case lenSize: // empty line
		return length, nil
	}

	if len(p) < length {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, len(p))
	}

	dataLen := length - lenSize
	dn, err := r.Read(p[:dataLen])
	if err != nil {
		return Err, err
	}

	if dn != dataLen {
		return Err, fmt.Errorf("%w: %d", ErrInvalidPktLen, dn)
	}

	if bytes.HasPrefix(p, errPrefix) {
		err = &ErrorLine{
			Text: string(bytes.TrimSpace(p[4:])),
		}
	}

	return length, err
}
```

This results in zero allocs and a _slight_ improvement over the current `Scanner` implementation:
```
› go test -bench=. -benchmem
goos: darwin
goarch: amd64
pkg: github.com/go-git/go-git/v5/plumbing/format/pktline
cpu: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
BenchmarkScanner/empty-16       90463454                13.10 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/one_message-16                 38528110                28.36 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/two_messages-16                41303529                28.42 ns/op            0 B/op          0 allocs/op
BenchmarkScanner/sections-16                    41827941                28.50 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/empty-16                    100000000               10.08 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/one_message-16              45403502                25.11 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/two_messages-16             43133918                25.16 ns/op            0 B/op          0 allocs/op
BenchmarkReadPacket/sections-16                 42402376                25.03 ns/op            0 B/op          0 allocs/op
BenchmarkEncoder/empty-16                       33795769                36.49 ns/op           19 B/op          1 allocs/op
BenchmarkEncoder/one_message-16                 27409754                43.34 ns/op           43 B/op          1 allocs/op
BenchmarkEncoder/two_messages-16                33945912                49.15 ns/op           67 B/op          1 allocs/op
BenchmarkEncoder/sections-16                    16685656               132.9 ns/op           261 B/op          1 allocs/op
BenchmarkWritePacket/empty-16                   94312168                12.10 ns/op           11 B/op          0 allocs/op
BenchmarkWritePacket/one_message-16             29564037                34.53 ns/op           40 B/op          1 allocs/op
BenchmarkWritePacket/two_messages-16            35636349                39.47 ns/op           64 B/op          1 allocs/op
BenchmarkWritePacket/sections-16                17210281                67.78 ns/op          253 B/op          1 allocs/op
PASS
ok      github.com/go-git/go-git/v5/plumbing/format/pktline     25.918s
```

You can find the changes I made to generate the benchmarks in this commit https://github.com/aymanbagabas/go-git/commit/33dc6724c4edbbd88a6c46a3e2b5ab730061d422